### ISSUE TITLE
API for data imports

### ISSUE THREAD

**Issue by mholt** [timelinize/timelinize#31]
Timelinize could benefit from an API to import data, or at least add items to the DB (both singly and in bulk).

This would make push-style ingestion possible.

**Comment by rubiojr** [comment]
This would be pretty nice IMO.

Somewhat related, it'd be pretty cool if Timelinize could eventually also accept Items with an unknown item class/type, and maybe render the metadata in a table, given it may not know how to render them graphically (like MP3 ID3v2 for example, or random CSV data with column names)

**Comment by mholt** [comment]
> Somewhat related, it'd be pretty cool if Timelinize could eventually also accept Items with an unknown item class/type,

I believe this is already possible. I've tried to make Timelinize as flexible as possible when ingesting data. It will try to infer missing info when needed; for example, it can guess the media type if omitted by the data source, and it can sometimes guess the item class. But yeah, data sources don't *have* to specify it (it's just recommended and useful).

> and maybe render the metadata in a table, given it may not know how to render them graphically (like MP3 ID3v2 for example, or random CSV data with column names)

If you go to the item view (the page that shows a single item) we do currently show metadata in key-value pairs as a "data grid" type of layout. Is that kind of what you mean?

**Comment by rubiojr** [comment]
> If you go to the item view (the page that shows a single item) we do currently show metadata in key-value pairs as a "data grid" type of layout. Is that kind of what you mean?

Yup. What kind of metadata is supposed to be rendered there? whatever `Metadata` the data source is adding in `FileImport`? So data source dependent? (that seems to be the case after looking at some data sources.

**Comment by mholt** [comment]
Yep, exactly. Right now it just dumps the key values of the metadata field provided by the data source.

In the future we can be smart about some metadata fields and show them in a more useful way but at least for now they're all displayed.

**Comment by mholt** [comment]
The current API for `/api/import` involves a payload containing this information:

- Name of data source that should perform the import
- List of filenames
- Processing options (these apply to all data sources and are mainly enforced by the processor, but for more efficient importing, data sources may also enforce certain processing options like timeframe)
- Data source options (these are specific to the data source being used)

In some sense, this API can already be called from a program on the loopback interface to "push" data to Timelinize. However...

The import flow is being revised, and as part of that, an import payload will look something more like this (subject to change still!):

- Processing options
- An array of:
    - Data source
    - List of filenames
    - Data source options

In the new flow, not all data needs to be ingested via an "import" -- a term which I'm soon going to reserve to a group or batch of data being ingested all at once. Items can still be added without being associated with an "import", such as items that are manually entered, or I guess, pushed!

So that leads me to wonder if the API for ingesting data should be generalized, or if we should have separate import/push/manual-entry APIs.

Pushing data via an HTTP request is interesting as it would involve sending potentially large, multipart payloads. Manual entry and push are probably very similar in fact.

Maybe a significant distinguishing factor is whether a "data source" (provider) is used or not. Data sources are, of course, packages that are hard-coded into the application (currently -- we can make this pluggable much like [Caddy plugins](https://caddyserver.com) but that's another topic). In other words, you give the name of the data source and the list of filenames to import along with any special options. The data source knows how to open the files and read them to make items, relationships, entities, etc.

Basically, data sources convert files (or folders) into [`Graph` values](https://pkg.go.dev/github.com/timelinize/timelinize@v0.0.0-20240822200625-a24d5758832c/timeline#Graph) which it sends down a channel for the processor.

But the push and manual-entry model is interesting because it doesn't really involve the use of pre-coded data sources. Instead, Graphs would be uploaded to Timelinize. Instead of giving Timelinize a filename, you give it the actual graph of items, relationships, entities, etc.

So it's a slightly different layer of abstraction and I wonder if a separate API would be more suitable.

The question is how to represent a stream of Graph objects in an HTTP request.

Would this be a multipart request or a stream of JSON objects that the server reads from until the client closes the connection? Or does the client just rapid-fire requests to the server with bits and pieces of a larger whole, like one Graph at a time? Ideally the server won't have to keep state between "pushes" to group related data together.

**Comment by rubiojr** [comment]
> In some sense, this API can already be called from a program on the loopback interface to "push" data to Timelinize. However...

Indeed, I've been using already to create some python scripts and seed new Timelinize timelines, speeding up the manual testing process when starting with an empty timeline. Working great so far.

> such as items that are manually entered, or I guess, pushed!.

Yep, this would be a great addition. Right now you need to have a file around and Timelinize needs to have access to its filesystem.

> So that leads me to wonder if the API for ingesting data should be generalized, or if we should have separate import/push/manual-entry APIs.
> Maybe a significant distinguishing factor is whether a "data source" (provider) is used or not. Data sources are, of course, packages that are hard-coded into the application (currently -- we can make this pluggable much like [Caddy plugins](https://caddyserver.com/) but that's another topic). In other words, you give the name of the data source and the list of filenames to import along with any special options. The data source knows how to open the files and read them to make items, relationships, entities, etc.

Good questions. The way I see it, it'd be nice to be able to push items via API, so if I host Timelinize remotely, I don't need to upload the file/directory first then ask Timelinize to scan/import it. At least for items that do not have an associated file and are stored in database. Although there's no reason why we shouldn't be able to push items with an associated attachment/file, [Pocketbase](https://github.com/pocketbase/pocketbase) files API comes to mind here. So maybe an optimized API for large files and then a way to associate (lightweigt) items pushed to them. Maybe keeping the existing datasources API and then a dedicated items and files API would work OK, without having to figure out how to cram everything into a unified API. Thinking out loud mostly.

**Comment by mholt** [comment]
A few updates on my thoughts about the import API:

There will need to be 2 ways to ingest data from external sources.

One is to have external data sources be able to push data one-by-one as they occur or are generated; i.e. no batch import job. Just one-off uploads to the timeline. This will be useful for clients which can generate data as time goes on. For example a GPS tracker or HomeAssistant (#21). As you move, it can push data points one by one to the timeline.

The other way is to have external data sources push batches of data as an import job. Since this issue was opened, I've already made data sources dynamic (they are no longer hard-coded), so third party scripts and such which have access to push to the timeline can declare their data source information.

(I still need to figure out things like authentication later; but for right now, we're just using a locally-bound, trusted socket during development.)
